src/main/java/io/cdap/plugin/cdc/source/salesforce/SalesforceReceiver.java

```diff
@@ -0,0 +1,250 @@
+ * is processed for the first time.
+ */
+public class SalesforceReceiver extends Receiver<StructuredRecord> {
+  private static final Logger LOG = LoggerFactory.getLogger(SalesforceReceiver.class);
+  private static final String RECEIVER_THREAD_NAME = "salesforce_streaming_api_listener";
+  // every x seconds thread wakes up and checks if stream is not yet stopped
+  private static final long GET_MESSAGE_TIMEOUT_SECONDS = 2;
+  private static final Gson GSON = new Gson();
+
+  private final AuthenticatorCredentials credentials;
+  private final List<String> objectsForTracking;
+  private final ErrorHandling errorHandling;
+  private final Map<String, Schema> schemas = new HashMap<>();
+  private final Map<String, List<ChangeEventHeader>> events = new HashMap<>();
+  private SalesforceEventTopicListener eventTopicListener;
+  private static final JsonParser JSON_PARSER = new JsonParser();
+
+  SalesforceReceiver(AuthenticatorCredentials credentials, List<String> objectsForTracking,
+                     ErrorHandling errorHandling) {
+    super(StorageLevel.MEMORY_AND_DISK_2());
+    this.credentials = credentials;
+    this.objectsForTracking = new ArrayList<>(objectsForTracking);
+    this.errorHandling = errorHandling;
+  }
+
+  @Override
+  public void onStart() {
+    eventTopicListener = new SalesforceEventTopicListener(credentials, objectsForTracking);
+    eventTopicListener.start();
+
+    ThreadFactory namedThreadFactory = new ThreadFactoryBuilder()
+      .setNameFormat(RECEIVER_THREAD_NAME + "-%d")
+      .build();
+
+    Executors.newSingleThreadExecutor(namedThreadFactory).submit(this::receive);
+  }
+
+  @Override
+  public void onStop() {
+    // There is nothing we can do here as the thread calling receive()
+    // is designed to stop by itself if isStopped() returns false
+  }
+
+  private void receive() {
+    PartnerConnection connection;
+    try {
+      connection = SalesforceConnectionUtil.getPartnerConnection(credentials);
+    } catch (ConnectionException e) {
+      throw new RuntimeException("Failed to connect to Salesforce", e);
+    }
+
+    while (!isStopped()) {
+      try {
+        String message = eventTopicListener.getMessage(GET_MESSAGE_TIMEOUT_SECONDS, TimeUnit.SECONDS);
+
+        if (message != null) {
+          // whole message class is not needed because we are interested only in change event payload
+          JsonObject headerElement = JSON_PARSER.parse(message)
+            .getAsJsonObject()
+            .getAsJsonObject("data")
+            .getAsJsonObject("payload")
+            .getAsJsonObject("ChangeEventHeader");
+          ChangeEventHeader event = GSON.fromJson(headerElement, ChangeEventHeader.class);
+
+          List<ChangeEventHeader> eventsList = events.getOrDefault(event.getTransactionKey(), new ArrayList<>());
+          eventsList.add(event);
+
+          if (event.isTransactionEnd()) {
```

Comment:
So does this mean the API returns events for different transactions interspersed with each other? For example:

tx1, change1
tx2, change1
tx1, change2

Do events for the same transaction have the same replay id or does each event has its own replay id?

Reply:
One transaction can create several change events. For example:
tx1, change1
tx1, change2
All change events related to one transaction will have same transaction key.